File: Methods-C++/sampleWeightedVariance.cpp
#include <iostream>
#include <math.h>

double sampleWeightedVariance(double data[], double weights[], int n) {
    double sumOfWeights = 0;
    double sumOfProducts = 0;
    double sumOfSquaredProducts = 0;
    for (int i = 0; i < n; i++) {
        sumOfWeights += weights[i];
        sumOfProducts += data[i] * weights[i];
        sumOfSquaredProducts += data[i] * data[i] * weights[i];
    }
    return (sumOfSquaredProducts - sumOfProducts * sumOfProducts / sumOfWeights) / (sumOfWeights - 1);
}

File: Methods-Python/sampleWeightedVariance.py
def sampleWeightedVariance(data, weights):

    sumOfWeights = 0
    sumOfProducts = 0
    sumOfSquaredProducts = 0

    for i in range(0, len(data)):
        sumOfWeights += weights[i]
        sumOfProducts += data[i] * weights[i]
        sumOfSquaredProducts += data[i] * data[i] * weights[i]

    return (sumOfSquaredProducts - sumOfProducts * sumOfProducts / sumOfWeights) / (sumOfWeights - 1)
